File: typescript/src/policyCertification.ts
/**
 * Policy Certification Lifecycle.
 *
 * Wire surface: `v1-certifications` edge function (POST only).
 * Supported operations: `list_policy_versions`, `create_approval`,
 * `list_attestations`.
 *
 * The certification flow:
 * 1. A policy version is published — a `PolicyVersion` row is created.
 * 2. Designated certifiers submit `PolicyApproval` records via
 *    `create_approval`.
 * 3. Once the quorum requirement is met the version transitions to
 *    `certified`; the attestation trail is readable via
 *    `list_attestations`.
 *
 * Wire-stable as `policy_certification.v1`.
 */

/** Lifecycle status of a policy version under certification. */
export type PolicyVersionStatus =
  | "draft"       // not yet submitted for approval
  | "pending"     // submitted; awaiting approvals
  | "certified"   // quorum met; policy is live
  | "rejected"    // blocked by a reject vote
  | "superseded"  // replaced by a newer certified version
  | "archived";   // removed from active use

/** A versioned snapshot of an org's policy. */
export interface PolicyVersion {
  readonly version_id: string;
  readonly org_id: string;
  /** Human-readable policy name. */
  readonly policy_name: string;
  /** Monotonically-increasing integer per `policy_name`. */
  readonly version_number: number;
  readonly status: PolicyVersionStatus;
  /** SHA-256 hex of the policy body. */
  readonly body_hash: string;
  readonly submitted_by: string;
  readonly submitted_at: string;
  /** ISO 8601 when the last approval quorum was met (null while pending). */
  readonly certified_at: string | null;
  /** Number of approvals received so far. */
  readonly approval_count: number;
  /** Minimum approvals required for certification. */
  readonly approval_quorum: number;
}

/** Response for `list_policy_versions`. */
export interface ListPolicyVersionsResponse {
  readonly versions: readonly PolicyVersion[];
  readonly total: number;
  readonly next_cursor?: string;
}

/** Query parameters for listing policy versions. */
export interface ListPolicyVersionsQuery {
  readonly policy_name?: string;
  readonly status?: PolicyVersionStatus;
  readonly limit?: number;
  readonly cursor?: string;
}

/** A single certifier approval record. */
export interface PolicyApproval {
  readonly approval_id: string;
  readonly version_id: string;
  readonly org_id: string;
  readonly approver_id: string;
  readonly approver_label: string;
  readonly decision: "approve" | "reject";
  readonly comment: string | null;
  readonly created_at: string;
}

/** Request body for `create_approval`. */
export interface CreatePolicyApprovalRequest {
  readonly version_id: string;
  readonly approver_id: string;
  readonly approver_label: string;
  readonly decision: "approve" | "reject";
  readonly comment?: string;
}

/** Response for `create_approval`. */
export interface CreatePolicyApprovalResponse {
  /** The recorded approval. */
  readonly approval: PolicyApproval;
  /** Updated policy version (reflects new `approval_count` / `status`). */
  readonly version: PolicyVersion;
}

/**
 * A signed attestation record emitted when a version reaches `certified`
 * status.  Attestations are append-only and immutable.
 */
export interface PolicyAttestation {
  readonly attestation_id: string;
  readonly version_id: string;
  readonly org_id: string;
  readonly policy_name: string;
  readonly version_number: number;
  readonly body_hash: string;
  readonly certified_at: string;
  /** SHA-256 over canonical JSON of the approval records. */
  readonly approval_chain_hash: string;
  /** Ordered list of approver IDs whose votes satisfied the quorum. */
  readonly approver_ids: readonly string[];
}

/** Response for `list_attestations`. */
export interface ListPolicyAttestationsResponse {
  readonly attestations: readonly PolicyAttestation[];
  readonly total: number;
  readonly next_cursor?: string;
}

/** Query parameters for listing attestations. */
export interface ListPolicyAttestationsQuery {
  readonly policy_name?: string;
  readonly limit?: number;
  readonly cursor?: string;
}

File: python/atlasent/policy_certification.py
"""Policy Certification Lifecycle wire types.

Parity with ``typescript/src/policyCertification.ts``.

Wire surface: ``v1-certifications`` edge function (POST only).
Operations: ``list_policy_versions``, ``create_approval``,
``list_attestations``.

Certification flow::

    1. A policy version is published → ``PolicyVersion`` row created
       with ``status="pending"``.
    2. Designated certifiers submit ``PolicyApproval`` records via
       ``create_approval``.
    3. Once the quorum requirement is met the version transitions to
       ``status="certified"`` and a ``PolicyAttestation`` is recorded.
    4. Attestations are append-only and immutable; read them via
       ``list_attestations``.

Wire-stable as ``policy_certification.v1``.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Literals
# ---------------------------------------------------------------------------

PolicyVersionStatus = Literal[
    "draft",
    "pending",
    "certified",
    "rejected",
    "superseded",
    "archived",
]
"""Lifecycle status of a policy version under certification."""

PolicyApprovalDecision = Literal["approve", "reject"]
"""Vote a certifier may cast on a policy version."""


# ---------------------------------------------------------------------------
# Models
# ---------------------------------------------------------------------------


class PolicyVersion(BaseModel):
    """A versioned snapshot of an org's policy."""

    version_id: str
    org_id: str
    policy_name: str
    version_number: int
    """Monotonically-increasing integer per ``policy_name``."""
    status: PolicyVersionStatus
    body_hash: str
    """SHA-256 hex of the policy body."""
    submitted_by: str
    submitted_at: str
    certified_at: str | None = None
    """ISO 8601 when the last approval quorum was met; ``None`` while pending."""
    approval_count: int = 0
    approval_quorum: int = 1

    model_config = {"extra": "allow"}


class ListPolicyVersionsResponse(BaseModel):
    """Response for ``list_policy_versions``."""

    versions: list[PolicyVersion] = Field(default_factory=list)
    total: int = 0
    next_cursor: str | None = None

    model_config = {"extra": "allow"}


class PolicyApproval(BaseModel):
    """A single certifier approval record."""

    approval_id: str
    version_id: str
    org_id: str
    approver_id: str
    approver_label: str
    decision: PolicyApprovalDecision
    comment: str | None = None
    created_at: str

    model_config = {"extra": "allow"}


class CreatePolicyApprovalRequest(BaseModel):
    """Request body for ``create_approval``."""

    version_id: str
    approver_id: str
    approver_label: str
    decision: PolicyApprovalDecision
    comment: str | None = None

    model_config = {"extra": "allow"}


class CreatePolicyApprovalResponse(BaseModel):
    """Response for ``create_approval``.

    Contains both the recorded approval and the updated policy version
    (reflects new ``approval_count`` / ``status``).
    """

    approval: PolicyApproval
    version: PolicyVersion

    model_config = {"extra": "allow"}


class PolicyAttestation(BaseModel):
    """Signed attestation record emitted when a version reaches ``certified``.

    Attestations are append-only and immutable.
    """

    attestation_id: str
    version_id: str
    org_id: str
    policy_name: str
    version_number: int
    body_hash: str
    certified_at: str
    approval_chain_hash: str
    """SHA-256 over canonical JSON of the approval records."""
    approver_ids: list[str] = Field(default_factory=list)
    """Ordered list of approver IDs whose votes satisfied the quorum."""

    model_config = {"extra": "allow"}


class ListPolicyAttestationsResponse(BaseModel):
    """Response for ``list_attestations``."""

    attestations: list[PolicyAttestation] = Field(default_factory=list)
    total: int = 0
    next_cursor: str | None = None

    model_config = {"extra": "allow"}


__all__ = [
    "CreatePolicyApprovalRequest",
    "CreatePolicyApprovalResponse",
    "ListPolicyAttestationsResponse",
    "ListPolicyVersionsResponse",
    "PolicyApproval",
    "PolicyApprovalDecision",
    "PolicyAttestation",
    "PolicyVersion",
    "PolicyVersionStatus",
]
